src/bootstrap/bootstrap.py

```diff
@@ -612,6 +631,30 @@ def fix_bin_or_dylib(self, fname, rpath_libz=False):
             print("warning: failed to call patchelf:", reason)
             return
 
+    # Return the stage1 compiler to download, if any.
+    def maybe_download_rustc(self):
+        # If `download-rustc` is not set, default to rebuilding.
+        if self.get_toml("download-rustc", section="rust") != "true":
+            return None
+
+        # Handle running from a directory other than the top level
+        rev_parse = ["git", "rev-parse", "--show-toplevel"]
+        top_level = subprocess.check_output(rev_parse, universal_newlines=True).strip()
+        compiler = "{}/compiler/".format(top_level)
+
+        # Look for a version to compare to based on the current commit.
+        # Ideally this would just use `merge-base`, but on beta and stable branches that wouldn't
+        # come up with any commits, so hack it and use `author=bors` instead.
+        merge_base = ["git", "log", "--author=bors", "--pretty=%H", "-n1", "--", compiler]
```

Comment:
This already went wrong :/ I just rebased over 9a1d6174c925f54c923599e29b09d6855e6b3a78, which is a bors commit, but git doesn't recognize as modifying any code:

```
$ git show 9a1d6174c925f54c923599e29b09d6855e6b3a78
error: refs/bisect/bad does not point to a valid object!
commit 9a1d6174c925f54c923599e29b09d6855e6b3a78 (origin/master, origin/HEAD, master)
Merge: 37d067f5d71 11e7897eae7
Author: bors <bors@rust-lang.org>
Date:   Sat Feb 6 18:03:37 2021 +0000

    Auto merge of #81832 - jonas-schievink:rollup-3nw53p0, r=jonas-schievink
    
    Rollup of 7 pull requests
    
    Successful merges:
    
     - #81402 (tidy: Run tidy style against markdown files.)
     - #81434 (BTree: fix documentation of unstable public members)
     - #81680 (Refactor `PrimitiveTypeTable` for Clippy)
     - #81737 (typeck: Emit structured suggestions for tuple struct syntax)
     - #81738 (Miscellaneous small diagnostics cleanup)
     - #81766 (Enable 'task list' markdown extension)
     - #81812 (Add a test for escaping LLVMisms in inline asm)
    
    Failed merges:
    
    r? `@ghost`
    `@rustbot` modify labels: rollup
$
```
`merge-base` correctly says 9a1d6174c925f54c923599e29b09d6855e6b3a78 was most recent change, but `log --author=bors` shows cfba499271ba53190a1d3647ff8f7202ec9ed6f5 instead, which is wrong, because compiler/ was modified since then (specifically in 7acf9ecf4f0c9e53f8a69c404b68509ae56e105a, 85fb5cdf2633e91c5101dd8fb89a637c6d007802, f631410159d99280c98e2df6db89a61f73d8b541, and 96e843ce6ae42e0aa519ba45e148269de347fd8). In this case it didn't hurt anything, but if any of the APIs used by rustdoc had changed, it would have caused a build failure.

I really do think merge-base is more correct here. But I'm ok with fixing it in a follow-up if you like.

Reply:
It should be possible to pass -m1 or something to log to treat merge commits as modification; it shouldn't be a fundamental mitigation.

Reply:
I tried `-m` and it didn't help. But I agree this can wait for a follow-up :)

Reply:
Another alternative is to run `git log --author=bors` without adding `compiler/` to the end, and then run `git log <output of previous command> compiler/` on top of that. But that doesn't seem any simpler than merge-base.